File: scm-java/src/main/scheme/Primitive.java
package scheme;

import java.math.BigInteger;

public abstract class Primitive {
    public abstract String name();
    public abstract String info();
    public abstract Object apply(SourcePos pos, Object[] arguments);

    protected void checkArgs(SourcePos pos, Object[] arguments, int minArgs, int maxArgs) {
        if (minArgs == maxArgs && arguments.length != minArgs) {
            throw new SchemeError(
                pos,
                "~s needs ~s arguments, but got ~s",
                name(), minArgs, arguments.length
            );
        }
        if (maxArgs == -1 && arguments.length < minArgs) {
            throw new SchemeError(
                pos,
                "~s needs at least ~s arguments, but got ~s",
                name(), minArgs, arguments.length
            );
        }
        if (maxArgs != -1 && (arguments.length < minArgs || arguments.length > maxArgs)) {
            throw new SchemeError(
                pos,
                "~s needs between ~s and ~s arguments, but got ~s",
                name(), minArgs, maxArgs, arguments.length
            );
        }
    }

    @Override
    public String toString()
    {
        return "#<p:" + name() + ">";
    }

    protected double toReal(Object value) {
        if (value instanceof Long) return (double) (long) (Long) value;
        if (Value.isBigInteger(value)) return IntegerMath.toDouble(value);
        if (value instanceof Rational) return ((Rational) value).toDouble();
        return (double) (Double) value;
    }

    protected boolean allIntegers(Object[] args) {
        for (Object arg : args) {
            if (!Value.isInteger(arg)) {
                return false;
            }
        }
        return true;
    }

    protected boolean allExactNums(Object[] args) {
        for (Object arg : args) {
            if (!Value.isInteger(arg) && !Value.isRational(arg))
                return false;
        }
        return true;
    }

    protected boolean hasComplex(Object[] args) {
        for (Object arg : args) {
            if (arg instanceof Complex) return true;
        }
        return false;
    }

    protected Object toIntegerIfPossible(Object value) {
        if (Value.isInteger(value)) return value;
        double val = toReal(value);
        if (val == Math.floor(val)) return (long) val;
        return value;
    }

    /**
     * Compare mixed exact/inexact numbers without precision loss.
     * Converts the inexact value to exact (BigInteger) when it's integral,
     * avoiding the double->long cast that loses precision beyond 2^53.
     */
    public static boolean mixedNumericEquals(Object a, Object b) {
        double d; Object exact;
        if (Value.isReal(a) && Value.isInteger(b)) { d = Value.asReal(a); exact = b; }
        else if (Value.isInteger(a) && Value.isReal(b)) { d = Value.asReal(b); exact = a; }
        else { return asDouble(a) == asDouble(b); }

        if (Double.isNaN(d) || Double.isInfinite(d)) return false;
        if (d != Math.floor(d)) return false; // non-integral double != integer
        // Convert double to BigInteger for precise comparison
        return IntegerMath.genericEquals(new java.math.BigDecimal(d).toBigInteger(), exact);
    }

    /**
     * Compare mixed exact/inexact for ordering (< or >).
     * Returns negative if a < b, positive if a > b, 0 if equal.
     */
    public static int mixedNumericCompare(Object a, Object b) {
        double d; Object exact; boolean flipped;
        if (Value.isReal(a) && Value.isInteger(b)) { d = Value.asReal(a); exact = b; flipped = false; }
        else if (Value.isInteger(a) && Value.isReal(b)) { d = Value.asReal(b); exact = a; flipped = true; }
        else { double da = asDouble(a), db = asDouble(b); return da < db ? -1 : da > db ? 1 : 0; }

        if (Double.isNaN(d)) return 0; // NaN comparisons are unordered
        if (d == Double.POSITIVE_INFINITY) return flipped ? -1 : 1;
        if (d == Double.NEGATIVE_INFINITY) return flipped ? 1 : -1;

        // Convert double to BigInteger for precise comparison
        BigInteger bi = new java.math.BigDecimal(d).toBigInteger();
        double trunc = d < 0 ? Math.ceil(d) : Math.floor(d);
        double frac = d - trunc;

        int cmp;
        if (exact instanceof Long) cmp = bi.compareTo(BigInteger.valueOf((long)(Long) exact));
        else cmp = bi.compareTo((BigInteger) exact);

        // If integer parts equal but double has fractional part, the double is larger/smaller
        if (cmp == 0 && frac != 0.0) cmp = frac > 0 ? 1 : -1;

        return flipped ? -cmp : cmp;
    }

    private static double asDouble(Object a) {
        if (a instanceof Double) return (double)(Double) a;
        if (a instanceof Long) return (double)(long)(Long) a;
        if (a instanceof BigInteger) return ((BigInteger) a).doubleValue();
        if (a instanceof Rational) return asDouble(((Rational) a).numerator) / asDouble(((Rational) a).denominator);
        return 0.0;
    }
}

File: scm-csharp/scheme/Primitive.cs
namespace scheme;

public abstract class Primitive
{
    public abstract string Name();
    public abstract string Info();
    public abstract object Apply(SourcePos? pos, object[] arguments);

    protected void CheckArgs(SourcePos? pos, object[] arguments, int minArgs, int maxArgs)
    {
        if (minArgs == maxArgs && arguments.Length != minArgs)
        {
            throw new SchemeError(
                pos,
                "~s needs ~s arguments, but got ~s",
                Name(), minArgs, arguments.Length
            );
        }
        if (maxArgs == -1 && arguments.Length < minArgs)
        {
            throw new SchemeError(
                pos,
                "~s needs at least ~s arguments, but got ~s",
                Name(), minArgs, arguments.Length
            );
        }
        if (maxArgs != -1 && (arguments.Length < minArgs || arguments.Length > maxArgs))
        {
            throw new SchemeError(
                pos,
                "~s needs between ~s and ~s arguments, but got ~s",
                Name(), minArgs, maxArgs, arguments.Length
            );
        }
    }
    
    public override string ToString()
    {
        return "#<p:" + Name() + ">";
    }

    protected double ToReal(object value)
    {
        if (value is long l) return (double)l;
        if (Value.IsBigInteger(value)) return IntegerMath.ToDouble(value);
        if (value is Rational r) return r.ToDouble();
        return (double)value;
    }

    protected bool AllIntegers(object[] args)
    {
        foreach (object arg in args)
        {
            if (!Value.IsInteger(arg))
            {
                return false;
            }
        }
        return true;
    }

    protected bool AllExactNums(object[] args)
    {
        foreach (object arg in args)
        {
            if (!Value.IsInteger(arg) && !Value.IsRational(arg))
                return false;
        }
        return true;
    }

    protected bool HasComplex(object[] args)
    {
        foreach (object arg in args)
        {
            if (arg is Complex)
                return true;
        }
        return false;
    }

    protected object ToIntegerIfPossible(object value)
    {
        if (Value.IsInteger(value)) return value;
        double val = ToReal(value);
        if (val == Math.Floor(val)) return (long) val;
        return value;
    }

    /// <summary>
    /// Compare mixed exact/inexact numbers without precision loss.
    /// Converts the inexact value to exact (BigInteger) when it's integral,
    /// avoiding the double→long cast that loses precision beyond 2^53.
    /// </summary>
    public static bool MixedNumericEquals(object a, object b)
    {
        double d; object exact;
        if (Value.IsReal(a) && Value.IsInteger(b)) { d = Value.AsReal(a); exact = b; }
        else if (Value.IsInteger(a) && Value.IsReal(b)) { d = Value.AsReal(b); exact = a; }
        else { return AsDouble(a) == AsDouble(b); }

        if (double.IsNaN(d) || double.IsInfinity(d)) return false;
        if (d != Math.Floor(d)) return false; // non-integral double ≠ integer
        // Convert double to BigInteger for precise comparison
        return IntegerMath.GenericEquals(new System.Numerics.BigInteger(d), exact);
    }

    /// <summary>
    /// Compare mixed exact/inexact for ordering (< or >).
    /// Returns negative if a < b, positive if a > b, 0 if equal.
    /// </summary>
    public static int MixedNumericCompare(object a, object b)
    {
        double d; object exact; bool flipped;
        if (Value.IsReal(a) && Value.IsInteger(b)) { d = Value.AsReal(a); exact = b; flipped = false; }
        else if (Value.IsInteger(a) && Value.IsReal(b)) { d = Value.AsReal(b); exact = a; flipped = true; }
        else { double da = AsDouble(a), db = AsDouble(b); return da < db ? -1 : da > db ? 1 : 0; }

        if (double.IsNaN(d)) return 0; // NaN comparisons are unordered
        if (double.IsPositiveInfinity(d)) return flipped ? -1 : 1;
        if (double.IsNegativeInfinity(d)) return flipped ? 1 : -1;

        // Convert double to BigInteger for precise comparison
        var bi = new System.Numerics.BigInteger(d);
        double frac = d - Math.Truncate(d);

        int cmp;
        if (exact is long el) cmp = bi.CompareTo(new System.Numerics.BigInteger(el));
        else cmp = bi.CompareTo((System.Numerics.BigInteger)exact);

        // If integer parts equal but double has fractional part, the double is larger/smaller
        if (cmp == 0 && frac != 0.0) cmp = frac > 0 ? 1 : -1;

        return flipped ? -cmp : cmp;
    }

    private static double AsDouble(object a)
    {
        if (a is double d) return d;
        if (a is long l) return (double)l;
        if (a is System.Numerics.BigInteger bi) return (double)bi;
        if (a is Rational r) return AsDouble(r.Numerator) / AsDouble(r.Denominator);
        return 0.0;
    }
}
